### ISSUE TITLE
Implement optimizer module

### ISSUE THREAD

**Issue by DmitrySoshnikov** [DmitrySoshnikov/regexp-tree#21]
Optimizer module should transform a regular expression into an _optimized_ version, replacing some sub-expressions with their idiomatic patterns. This might be good for different kinds of minifiers, as well as for regexp machines.

Example:

```js
/[a-zA-Z_0-9][a-zA-Z_0-9]*\e{1,}/
```

Is transformed into:

```js
/\w+e+/
```

We can implement the optimizer as a set of small transforms (this depends on [issue #7](https://github.com/DmitrySoshnikov/regexp-tree/issues/7), and [issue #23](https://github.com/DmitrySoshnikov/regexp-tree/issues/23)). 

**Patterns to replace:**

- `aa*` -> `a+`
- `()`, `(?:)` - `<empty>` (remove empty groups)
- `a{1,}` -> `a+`
- `a{1}` -> `a`
- `a{3,3}` -> `a{3}`
- `[a-zA-Z_0-9]` -> `\w`
- `[^a-zA-Z_0-9]` -> `\W`
- `[\d\d]` -> `[\d]` (remove duplicates from char-class)
- `[\d]` -> `\d` (remove char class for single char)
- `[0-9]` -> `\d`
- `[^0-9]` -> `\D`
- `[ \t\r\n\f]` -> `\s`
- `[^ \t\r\n\f]` -> `\S`
- `\e` -> `e` (remove unnecessary escape)
- ...

See [TODO item](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/index.js#L80), optimizer lives in [src/optimizer](https://github.com/DmitrySoshnikov/regexp-tree/tree/master/src/optimizer).

**Comment by mathiasbynens** [comment]
Keep in mind that the `u` flag (or [the combination of the `u` and `i` flag](https://mathiasbynens.be/notes/es6-unicode-regex#impact-i)) impacts some of these transformations.

**Comment by DmitrySoshnikov** [comment]
@mathiasbynens, yeah, good point. This should probably be done only if `u` is not set (unless there are better options).

**Comment by hg42** [comment]
would be nice if we could choose, which transformations should be done.
I think I would need both, an include-list and an exclude-list.
The names should be have a scheme, e.g. all optimizations with quantifier could be called quantifier-xxx, or you could have xxx-reduce-xxx.

Optimizations are a special variant of a group of transformations.

IMHO, an optimization would be a group of transformations with an intent, e.g. make the expression smaller or remove illegal constructs or remove character classes or remove ranges where possible. There could be groups of transformations for these intents:
```js
optimize_size = ["char_range_to_class_digit", "char_range_to_class_word", "char_range_to_class_space"];
fix_illegal_constructs = ["remove_empty_groups", "remove_unnecessary_escapes"];
...
transformations = optimize_size;
ast = regexpTree.transform(ast, transformations);
```
You could add or subtract such groups or filter them.

**Comment by DmitrySoshnikov** [comment]
Yeah, good idea, `optimize` can accept a set of explicit transforms. Although, this use-case is already covered by `transform` -- you may pass any transforms you need. The `optimized` was supposed to be a "black-box" which doesn't bother you with explicit transforms (for what `transform` exists). But I'm fine with accepting a needed list of transforms as well there -- if it's not passed, a default can be used.

Good point on naming. I'm opened to suggestions on standardized naming.

**Comment by hg42** [comment]
> use-case is already covered by transform -- you may pass any transforms you need

`transform()` has a handlers object parameter (same as `traverse()`).
But `transformations` above is essentially an array of handler objects.

These are referred to by name in my suggestion, to choose from an existing internal set, but probably directly passing handler arrays (or objects) would be more flexible. The user may want to build his own handlers and sets.

> accepting a needed list of transforms as well there -- if it's not passed, a default can be used

yepp

**Comment by DmitrySoshnikov** [comment]
`transform`/`traverse` are overloaded, and accept as a single handler object (for convince), as well as array of handler objects. So one could write:

```js
tree.transform('/a+/', {
  Char({node}) {
    ...
  }
});
```

Or (a list of transforms, 1 handler in the list):

```js
tree.transform('/a+/', [{
  Char({node}) {
    ...
  }
}]);
```

Multiple handlers (run in one traversal pass, and applied in order; if a node is removed, the next handled is not called for it):

```js
tree.transform('/a+/', [
  {
    Char({node}) {
      ...
    }
  },
  {
    Char({node}) {
      ...
    }
  },
]);
```

**Comment by hg42** [comment]
nice, I didn't notice that.
so it's there already.

While this optimization may be working well in many transforms, in my use-case this single pass behaviour doesn't work. I need multiple passes.

E.g. I set the names of unnamed capture groups to it's index while collecting the hierarchical dependencies (collecting names of all parents). Then in a second pass, I delete all names, to have only un-named groups.

I guess, there are several tree operations, especially restructuring operations, where a pass has to be completed before the next handlers can be applied.

Also, creating new nodes in one handler set will not guaranty that they are feeded to the next handler set if they are merged in this way. And removing nodes backwards doesn't hide them from the next set, because they are already processed.

Without merging, you can also write `transformations.forEach(...)`, so it is not necessary to do this inside transform.

**Comment by hg42** [comment]
I just had an idea for the handlers parameter of transform().
It could have a form
```
[ [a, b], [c, d, e], [f] ]
```
where the outer level would be sequential and the inner level merged.
Here a and b would be merged and executed, then c, d, e merged and executed, then f.
If it would be `[a, b, c]`, a, b, c would be merged (like it is now).

**Comment by DmitrySoshnikov** [comment]
> While this optimization may be working well in many transforms, in my use-case this single pass behaviour doesn't work. I need multiple passes.

Yep, for sure there might be some complex cases when a single pass is not enough, and nodes may collide, so the multi-pass is an alternative (slower, but still). The single pass is only an optimization, which might be enough for most of use-cases.

> Also, creating new nodes in one handler set will not guaranty that they are feeded to the next handler set if they are merged in this way. And removing nodes backwards doesn't hide them from the next set, because they are already processed.

Yes, if a node is inserted somewhere to further positions, it'll be visited actually. However, if if it's inserted into already passed position, it'll be up to a user to traverse this node manually (and for this we'll provide API on `NodePath` to traverse).

> `[ [a, b], [c, d, e], [f] ]`
where the outer level would be sequential and the inner level merged.

Yeah, potentially we'll be able to have a set of "presets" (similarly to your example), i.e. set of transforms which are guaranteed to be ran in one pass, and not collide.

**Comment by DmitrySoshnikov** [comment]
Closing this one since the module is set up. Other specific transforms can be implemented in separate issues.
